docs/python/tutorial-fastapi.md

```diff
@@ -0,0 +1,595 @@
+
+You should now be ready to move on to the next section, where we will replace the data storage.
+
+## Replace the database
+1. In the `main.py` file, replace the `grocery_list = {}` in the beginning of the file with the lines below:
+    ```python
+    redis_client = redis.StrictRedis(host='0.0.0.0', port=6379, db=0, decode_responses=True)
+    ```
+Pylance will display an error message because redis hasn’t been imported yet.
+
+2. Click on "redis" in the editor, and select the light bulb the shows up (or `kb(editor.action.quickFix)`). Then select "Add 'import redis'".
+    ![Light bulb displayed next to the redis variable, with the option to add the import statement.](images/fastapi-tutorial/fastapi_add_redis_quickfix.png)
+
+You can also set up Pylance to automatically add imports by looking for the "Auto Import Completions" setting in the Settings UI page (`kb(workbench.action.openSettings)`) and enabling it.
+
+We now have a Redis client object that connects to a Redis server running on the local host (`host="0.0.0.0"`) and listening on port 6379 (`port=6379`). The `db` parameter specifies the Redis database to use. Redis supports multiple databases, and in this code we’re setting to use database 0, which is the default database. We're also passing `decode_responses=True` for the responses to be decoded as strings (instead of bytes).
+
+This client can now be used to perform operations such as getting, setting, and deleting objects in the Redis database.
+
+Let's do some more replacements in the first route `add_item`. Instead of looking into all the keys from the dictionary to find the item name that has been provided, we can fetch that information directly from a redis hash. Let’s assume we will create a hash called `item_name_to_id`, with item names mapped into their IDs once they are added in the app. We can then get the ID of the item name we’re receiving by invoking the `hget` method from redis.
+
+3. Delete the line with the content below:
+    ```python
+    items_ids = {item.item_name: item.item_id if item.item_id is not None else 0 for item in grocery_list.values()}
+    ```
+    And replace it with:
+    ```python
+      item_id = redis_client.hget("item_name_to_id", item_name)
+     ```
+
+Note that Pylance raises a problem with this change. This is because the `hget` method returns either `str`, or `None` (if the item doesn’t exist). However, the lines below the code that we haven’t replaced yet expect `item_id` to be of type `int`. Let’s address it by renaming the `item_id` symbol.
+
+4. Rename `item_id` with `item_id_str` by selecting the symbol and running the **Rename Symbol** command (`kb(editor.action.rename)`).
+
+    This will rename all occurrences of the symbol in the file, including the one in the `return` statement.
+
+5. Pylance should show a variable type hint next to `item_id_str`. Double click to accept it:
```

Comment:
It doesn't because the `item_id` assignments in the `if`/`else` block just below the replaced line were also renamed. And on those lines `item_id_str` (used to be `item_id`) is already annotated, as an `int`.

If I remove those `int` annotations, I get the `str | None` inlay hint.

Reply:
ha, good point. I didn't have the type hints in the beginning when I tried the rename symbol flow.  To be consistent with the code sample I ended up casting item_id as an int in the beginning.